File: basicCalculator/basicCalculator.cpp
#include <iostream>
#include <stack>
using namespace std;

class Solution {
public:
    int calculate(string s) {
        int res = 0;
        stack<int> stk;
        int i = 0;
        int sign = 1;
        while (i < s.length()) {
            if (isdigit(s[i])) {
                int num = 0;
                while (i < s.length() && isdigit(s[i])) {
                    num = (s[i] - '0') + num * 10;
                    i++;
                }
                i--;
                res += sign * num;
            } else if (s[i] == '+') {
                sign = 1;
            } else if (s[i] == '-') {
                sign = -1;
            } else if (s[i] == '(') {
                stk.push(res);
                stk.push(sign);
                res = 0;
                sign = 1;
            } else if (s[i] == ')') {
                res *= stk.top();
                stk.pop();
                res += stk.top();
                stk.pop();
            }
            i++;
        }
        return res;
    }
};

int main() {
    Solution s;
    string str = "(1+(4+5+2)-3)+(6+8)";
    cout << s.calculate(str) << endl;
    return 0;
}

File: 224.basicCalculator/basicCalculator.java
class Solution {
    public int calculate(String s) {
        int res = 0;
        int number = 0;
        int sign = 1;
        Stack<Integer> stk = new Stack<>();

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);

            if (c == ' ') {
                continue;
            } else if (Character.isDigit(c)) {
                number = number * 10 + c - '0';
            } else if (c == '+') {
                res += sign * number;
                number = 0;
                sign = 1;
            } else if (c == '-') {
                res += sign * number;
                number = 0;
                sign = -1;
            } else if (c == '(') {
                // 1. save current context to stack
                stk.push(res);
                stk.push(sign);

                // 2. reset context
                res = 0;
                number = 0;
                sign = 1;
            } else if (c == ')') {
                // update result
                res += sign * number;
                number = 0;

                // recover outer context 
                int outerSign = stk.pop();
                int outerRes = stk.pop();

                res = outerRes + outerSign * res;
                sign = 1;
            }
        }

        // commit last
        res += sign * number;
        return res;
    }
}